File: extract_bionumerics.py
# =============================================================================
# Extract sequences and metadata from BioNumerics export
# =============================================================================

import os
import base64
import csv
import re
from xml.etree import ElementTree as ET
from pathlib import Path

# Set paths
export_dir = Path("export")
entries_dir = export_dir / "_Entries"
experiments_dir = export_dir / "_Experiments"
output_fasta = Path("data/bionumerics_sequences.fasta")
output_metadata = Path("data/bionumerics_metadata.csv")

def decode_b64(text):
    """Decode Base64 string to UTF-8 text"""
    if not text:
        return ""
    try:
        return base64.b64decode(text).decode('utf-8')
    except:
        return text

def get_text_decoded(element):
    """Get text content, decoding Base64 if needed"""
    if element is None:
        return ""
    text = element.text or ""
    if element.get("B64") == "1" and text:
        return decode_b64(text)
    return text

# -----------------------------------------------------------------------------
# Parse all Entry files to get metadata
# -----------------------------------------------------------------------------
print("Parsing metadata from _Entries...")

all_entries = []
entry_files = sorted(entries_dir.glob("*.xml"))

for ef in entry_files:
    print(f"  Reading: {ef.name}")
    tree = ET.parse(ef)
    root = tree.getroot()
    
    for entry in root.findall(".//Entry"):
        # Get sample key (ID)
        key_node = entry.find(".//Key")
        sample_id = get_text_decoded(key_node) if key_node is not None else ""
        
        # Get experiment path (links to sequence file)
        exp_path_node = entry.find(".//SEQ/Path")
        exp_path = get_text_decoded(exp_path_node) if exp_path_node is not None else ""
        
        # Get all fields
        entry_data = {
            "sample_id": sample_id,
            "exp_path": exp_path
        }
        
        for field in entry.findall(".//Field"):
            name_node = field.find("Name")
            value_node = field.find("Value")
            
            field_name = get_text_decoded(name_node) if name_node is not None else ""
            field_value = get_text_decoded(value_node) if value_node is not None else ""
            
            entry_data[field_name] = field_value
        
        all_entries.append(entry_data)

print(f"Found {len(all_entries)} entries")

# -----------------------------------------------------------------------------
# Parse Experiment files to get sequences
# -----------------------------------------------------------------------------
print("\nParsing sequences from _Experiments...")

sequences = {}
missing_files = 0

for i, entry in enumerate(all_entries):
    exp_path = entry.get("exp_path", "")
    sample_id = entry.get("sample_id", "")
    
    if not exp_path or not sample_id:
        continue
    
    # Convert Windows path separator
    exp_file = export_dir / exp_path.replace("\\", "/")
    
    if exp_file.exists():
        try:
            tree = ET.parse(exp_file)
            root = tree.getroot()
            seq_node = root.find(".//SequenceData")
            
            if seq_node is not None and seq_node.text:
                seq_data = seq_node.text.strip().upper()
                if seq_data:
                    sequences[sample_id] = seq_data
        except Exception as e:
            print(f"  Error reading {exp_file}: {e}")
    else:
        missing_files += 1
    
    if (i + 1) % 100 == 0:
        print(f"  Processed {i + 1} of {len(all_entries)}")

print(f"Found {len(sequences)} sequences")
if missing_files > 0:
    print(f"  ({missing_files} experiment files not found)")

# -----------------------------------------------------------------------------
# Create output directory
# -----------------------------------------------------------------------------
output_fasta.parent.mkdir(exist_ok=True)

# -----------------------------------------------------------------------------
# Create output metadata (only for samples with sequences)
# -----------------------------------------------------------------------------
print("\nCreating output files...")

def convert_date(date_str):
    """Convert DD.MM.YYYY to YYYY-MM-DD"""
    if not date_str:
        return ""
    match = re.match(r"(\d{2})\.(\d{2})\.(\d{4})", date_str)
    if match:
        day, month, year = match.groups()
        return f"{year}-{month}-{day}"
    return date_str

# Prepare metadata rows
metadata_rows = []
for entry in all_entries:
    sample_id = entry.get("sample_id", "")
    if sample_id not in sequences:
        continue
    
    row = {
        "sample_id": sample_id,
        "sampling_date": convert_date(entry.get("SAMPLE_DATE", "")),
        "sample_year": entry.get("SAMPLE_YEAR", ""),
        "genotype": entry.get("GENOTYPE", ""),
        "outbreak_variant": entry.get("OUTBREAK_VARIANT", ""),
        "origin": entry.get("ORIGIN", ""),
        "source": entry.get("SOURCE", ""),
        "transmission_route": entry.get("TRANSMISSION_ROUTE", ""),
        "patient_initials": entry.get("PATIENT_INITIALS", ""),
        "group": entry.get("GROUP", ""),
        "sequence_length": len(sequences[sample_id])
    }
    metadata_rows.append(row)

# Write metadata CSV
fieldnames = ["sample_id", "sampling_date", "sample_year", "genotype", 
              "outbreak_variant", "origin", "source", "transmission_route",
              "patient_initials", "group", "sequence_length"]

with open(output_metadata, "w", newline="", encoding="utf-8") as f:
    writer = csv.DictWriter(f, fieldnames=fieldnames)
    writer.writeheader()
    writer.writerows(metadata_rows)

print(f"Wrote metadata to: {output_metadata}")

# -----------------------------------------------------------------------------
# Write FASTA file
# -----------------------------------------------------------------------------
with open(output_fasta, "w", encoding="utf-8") as f:
    for sample_id, sequence in sequences.items():
        f.write(f">{sample_id}\n")
        f.write(f"{sequence}\n")

print(f"Wrote {len(sequences)} sequences to: {output_fasta}")

# -----------------------------------------------------------------------------
# Summary
# -----------------------------------------------------------------------------
print("\n" + "=" * 50)
print("SUMMARY")
print("=" * 50)
print(f"Total entries in database: {len(all_entries)}")
print(f"Sequences extracted: {len(sequences)}")
print(f"\nOutput files:")
print(f"  - FASTA: {output_fasta}")
print(f"  - Metadata: {output_metadata}")

# Genotype distribution
print("\nGenotype distribution:")
genotypes = {}
for row in metadata_rows:
    gt = row["genotype"] or "(empty)"
    genotypes[gt] = genotypes.get(gt, 0) + 1
for gt, count in sorted(genotypes.items()):
    print(f"  {gt}: {count}")

# Year distribution
print("\nYear distribution:")
years = {}
for row in metadata_rows:
    yr = row["sample_year"] or "(empty)"
    years[yr] = years.get(yr, 0) + 1
for yr, count in sorted(years.items()):
    print(f"  {yr}: {count}")

# Sequence length stats
lengths = [row["sequence_length"] for row in metadata_rows]
if lengths:
    print(f"\nSequence lengths:")
    print(f"  Min: {min(lengths)} bp")
    print(f"  Max: {max(lengths)} bp")
    print(f"  Mean: {sum(lengths) // len(lengths)} bp")

File: extract_bionumerics.R
# =============================================================================
# Extract sequences and metadata from BioNumerics export
# =============================================================================

library(xml2)
library(base64enc)
library(dplyr)
library(tidyr)

# Set paths
export_dir <- "export"
entries_dir <- file.path(export_dir, "_Entries")
experiments_dir <- file.path(export_dir, "_Experiments")
output_fasta <- "data/bionumerics_sequences.fasta"
output_metadata <- "data/bionumerics_metadata.csv"

# Helper function to decode Base64
decode_b64 <- function(x) {

  if (is.na(x) || x == "") return("")
  tryCatch({
    rawToChar(base64decode(x))
  }, error = function(e) x)
}

# -----------------------------------------------------------------------------
# Parse all Entry files to get metadata
# -----------------------------------------------------------------------------
cat("Parsing metadata from _Entries...\n")

entry_files <- list.files(entries_dir, pattern = "\\.xml$", full.names = TRUE)
all_entries <- list()

for (ef in entry_files) {
  cat("  Reading:", basename(ef), "\n")
  doc <- read_xml(ef)
  entries <- xml_find_all(doc, ".//Entry")
  
  for (entry in entries) {
    # Get sample key (ID)
    key_node <- xml_find_first(entry, ".//Key")
    key_b64 <- xml_attr(key_node, "B64")
    sample_id <- if (!is.na(key_b64) && key_b64 == "1") {
      decode_b64(xml_text(key_node))
    } else {
      xml_text(key_node)
    }
    
    # Get experiment path (links to sequence file)
    exp_path_node <- xml_find_first(entry, ".//SEQ/Path")
    exp_path <- ""
    if (!is.na(xml_text(exp_path_node))) {
      path_b64 <- xml_attr(exp_path_node, "B64")
      exp_path <- if (!is.na(path_b64) && path_b64 == "1") {
        decode_b64(xml_text(exp_path_node))
      } else {
        xml_text(exp_path_node)
      }
    }
    
    # Get all fields
    fields <- xml_find_all(entry, ".//Field")
    field_data <- list(sample_id = sample_id, exp_path = exp_path)
    
    for (field in fields) {
      name_node <- xml_find_first(field, ".//Name")
      value_node <- xml_find_first(field, ".//Value")
      
      # Decode field name
      name_b64 <- xml_attr(name_node, "B64")
      field_name <- if (!is.na(name_b64) && name_b64 == "1") {
        decode_b64(xml_text(name_node))
      } else {
        xml_text(name_node)
      }
      
      # Decode field value
      value_b64 <- xml_attr(value_node, "B64")
      field_value <- if (!is.na(value_b64) && value_b64 == "1") {
        decode_b64(xml_text(value_node))
      } else {
        xml_text(value_node)
      }
      
      field_data[[field_name]] <- field_value
    }
    
    all_entries[[length(all_entries) + 1]] <- field_data
  }
}

# Convert to data frame
metadata_df <- bind_rows(all_entries)
cat("Found", nrow(metadata_df), "entries\n")

# -----------------------------------------------------------------------------
# Parse Experiment files to get sequences
# -----------------------------------------------------------------------------
cat("\nParsing sequences from _Experiments...\n")

sequences <- list()

for (i in 1:nrow(metadata_df)) {
  exp_path <- metadata_df$exp_path[i]
  sample_id <- metadata_df$sample_id[i]
  
  if (is.na(exp_path) || exp_path == "") next
  
  # Convert Windows path to current OS path
  exp_file <- file.path(export_dir, gsub("\\\\", "/", exp_path))
  
  if (file.exists(exp_file)) {
    doc <- read_xml(exp_file)
    seq_node <- xml_find_first(doc, ".//SequenceData")
    
    if (!is.na(seq_node)) {
      seq_data <- xml_text(seq_node)
      if (!is.na(seq_data) && nchar(seq_data) > 0) {
        sequences[[sample_id]] <- toupper(seq_data)
      }
    }
  }
  
  if (i %% 50 == 0) cat("  Processed", i, "of", nrow(metadata_df), "\n")
}

cat("Found", length(sequences), "sequences\n")

# -----------------------------------------------------------------------------
# Create output metadata (only for samples with sequences)
# -----------------------------------------------------------------------------
cat("\nCreating output files...\n")

# Filter metadata to only include samples with sequences
metadata_output <- metadata_df %>%
  filter(sample_id %in% names(sequences)) %>%
  select(
    sample_id,
    sampling_date = SAMPLE_DATE,
    sample_year = SAMPLE_YEAR,
    genotype = GENOTYPE,
    outbreak_variant = OUTBREAK_VARIANT,
    origin = ORIGIN,
    source = SOURCE,
    transmission_route = TRANSMISSION_ROUTE,
    patient_initials = PATIENT_INITIALS,
    group = GROUP
  ) %>%
  mutate(
    # Convert date format from DD.MM.YYYY to YYYY-MM-DD
    sampling_date = case_when(
      grepl("^\\d{2}\\.\\d{2}\\.\\d{4}$", sampling_date) ~ {
        parts <- strsplit(sampling_date, "\\.")[[1]]
        paste(parts[3], parts[2], parts[1], sep = "-")
      },
      TRUE ~ sampling_date
    )
  )

# Fix date conversion (need to do row-wise)
metadata_output$sampling_date <- sapply(metadata_output$sampling_date, function(d) {
  if (is.na(d) || d == "") return(NA)
  if (grepl("^\\d{2}\\.\\d{2}\\.\\d{4}$", d)) {
    parts <- strsplit(d, "\\.")[[1]]
    return(paste(parts[3], parts[2], parts[1], sep = "-"))
  }
  return(d)
})

# Create data directory if needed
dir.create("data", showWarnings = FALSE)

# Write metadata CSV
write.csv(metadata_output, output_metadata, row.names = FALSE, na = "")
cat("Wrote metadata to:", output_metadata, "\n")

# -----------------------------------------------------------------------------
# Write FASTA file
# -----------------------------------------------------------------------------
fasta_lines <- c()
for (sample_id in names(sequences)) {
  fasta_lines <- c(fasta_lines, paste0(">", sample_id))
  fasta_lines <- c(fasta_lines, sequences[[sample_id]])
}

writeLines(fasta_lines, output_fasta)
cat("Wrote", length(sequences), "sequences to:", output_fasta, "\n")

# -----------------------------------------------------------------------------
# Summary
# -----------------------------------------------------------------------------
cat("\n=== Summary ===\n")
cat("Total entries in database:", nrow(metadata_df), "\n")
cat("Sequences extracted:", length(sequences), "\n")
cat("Output files:\n")
cat("  - FASTA:", output_fasta, "\n
")
cat("  - Metadata:", output_metadata, "\n")

# Show genotype distribution
cat("\nGenotype distribution:\n")
print(table(metadata_output$genotype, useNA = "ifany"))

cat("\nYear distribution:\n")
print(table(metadata_output$sample_year, useNA = "ifany"))
